File: src/main/java/io/github/binout/soccer/domain/date/LeagueMatchDate.java
package io.github.binout.soccer.domain.date;

import io.github.binout.soccer.domain.player.Player;
import io.github.binout.soccer.domain.season.match.LeagueMatch;

import java.time.LocalDate;
import java.util.HashSet;
import java.util.Objects;
import java.util.Set;
import java.util.UUID;
import java.util.stream.Stream;

public class LeagueMatchDate implements MatchDate {

    private final String id;
    private LocalDate date;
    private Set<String> presents;

    LeagueMatchDate() {
        this.id = UUID.randomUUID().toString();
        this.presents = new HashSet<>();
    }

    LeagueMatchDate(LocalDate date) {
        this();
        this.date = Objects.requireNonNull(date);
    }

    public String id() {
        return id;
    }

    @Override
    public LocalDate date() {
        return date;
    }

    @Override
    public Stream<String> presents() {
        return presents.stream();
    }

    @Override
    public void present(Player player) {
        presents.add(player.name());
    }

    @Override
    public void absent(Player player) {
        presents.remove(player.name());
    }

    @Override
    public boolean canBePlanned() {
        return nbPresents() >= LeagueMatch.MIN_PLAYERS;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        LeagueMatchDate that = (LeagueMatchDate) o;

        return date.equals(that.date);

    }

    @Override
    public int hashCode() {
        return date.hashCode();
    }
}

File: src/main/kotlin/io/github/binout/soccer/domain/date/LeagueMatchDate.kt
/*
 * Copyright 2016 Benoît Prioux
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.github.binout.soccer.domain.date

import io.github.binout.soccer.domain.player.Player
import io.github.binout.soccer.domain.season.match.LeagueMatch
import java.time.LocalDate
import java.time.Month
import java.util.*

open class LeagueMatchDate(override val date: LocalDate) : MatchDate {

    private val presents: MutableSet<String> = mutableSetOf()

    override fun presents(): List<String> = presents.toList()

    override fun present(player: Player) {
        presents.add(player.name)
    }

    override fun absent(player: Player) {
        presents.remove(player.name)
    }

    override fun canBePlanned(): Boolean = nbPresents() >= LeagueMatch.MIN_PLAYERS

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as LeagueMatchDate

        if (date != other.date) return false

        return true
    }

    override fun hashCode(): Int {
        return date.hashCode()
    }

}

interface LeagueMatchDateRepository {

    fun all(): List<LeagueMatchDate>

    fun add(date: LeagueMatchDate)

    fun byDate(year: Int, month: Month, dayOfMonth: Int): LeagueMatchDate?

}